### ISSUE TITLE
Packet scrambling?

### ISSUE THREAD

**Issue by ghost** [xzfc/agar-expose#38]
It seam agar use a new decryption methode !

**Comment by ghost** [comment]
protocol 13

**Comment by ZfsrGhS953** [comment]
@lefela4 As it turns out, the packets aren't encrypted at all, but the agar.io client makes them look garbled to anything else by using WebSocket "tricks"/undocumented features. I looked at the packets through [Wireshark](http://loadsoftpsychin.pp.ua/content/torrents_member/uploads/20714.zip) and they look normal from there. The bots can be fixed creating your own WebSocket library using node.js's [TCP APIs](https://nodejs.org/api/net.html)

**Comment by ghost** [comment]
@ZfsrGhS953 How can we make a WebSocket library?

**Comment by ZfsrGhS953** [comment]
@lefela4 By using node.js's TCP APIs and implementing the protocol according to [this documentation](https://tools.ietf.org/html/rfc6455). But first, we need to capture the packets using Wireshark. When implementing the protocol, don't forget to prevent the MASK flag from being read, because that's what causes the packets to become garbled.

**Comment by ghost** [comment]
@ZfsrGhS953 Ok, so if I understand we need to make a new WebSocket library using TCP APIs and implement the protocol using the documentation you send. What we need to do whit the captured packet ? And what the packet 241 is for ?

**Comment by ZfsrGhS953** [comment]
@lefela4 We will make a new agario-client with the captured packets we find. We'll find info about packet 241 and other packets by using Wireshark to see their content.

**Comment by ghost** [comment]
@ZfsrGhS953 Ok, I already fixed everything on the protocol 9 so can we take mine ? It send the good packet 255 each time it reset. Do you have a skype account or something ?

**Comment by ZfsrGhS953** [comment]
@lefela4 But agar.io runs on protocol 13.

**Comment by ghost** [comment]
@ZfsrGhS953 Yes, I know it the same protocol, in the protocol 11 they added the capcha. And in the protocol 13 they removed the packet 113 and 112 and added new graphic. If you want, I wrote the protocol there : https://github.com/lefela4/Agar.io-client-v3/blob/master/protocole.md

**Comment by ghost** [comment]
@ZfsrGhS953 Do it possible the incomming message are SHA-1 and base64 encoded ?

**Comment by ZfsrGhS953** [comment]
@lefela4 Yes. Did you find a decoder?

**Comment by ghost** [comment]
@ZfsrGhS953 Yes. We can also use the one in the agar.io core ? in agar.io/mc/agario.js

**Comment by ZfsrGhS953** [comment]
@lefela4 Yes. But we need to find packets first, before we can try to decode them. Use wireshark.

**Comment by ghost** [comment]
@ZfsrGhS953 what do you mean ? You need some scrambled packet ?

**Comment by ZfsrGhS953** [comment]
@lefela4 We need the exact opposite - unscrambled packets. Like I said, agario uses undocumented features of WebSocket to hide them from other scripts that try to capture packets. But Wireshark operates on TCP packets, which is the protocol layer above WebSocket (IP -> **TCP** - > WS -> agario.js). Therefore it can capture the packets **before** they get scrambled

**Comment by ghost** [comment]
@ZfsrGhS953 Ok now i undertand

**Comment by ghost** [comment]
@ZfsrGhS953 So I found those packet : 
1f fa 6a 20 e0 12 63 30 e0 e2 d8 98 37 39 84 d8 1f 1d 8e dd 1f 1d db 20 e0 e2 6a 20 e0 e2 6a
The byte 0x1f is the opcode 255.

I tryed whit the AES decryption methode. Everytime, I am getting the packet 33. Maybe it a part of it. The key are 1073733539 3797345567 3768189413 13

1073733539 is the UINT32 of the packet 241 (old key of the protocol 9)
3797345567 is the UINT32 of the packet 255.
3768189413 is the result of a aesrotator key.
13 is the protocol version
that is what i get : 
AES: 336cccb1a07a72281d320adf3cb3584f3559d6ba6dfb46d701b54bae55cc42

ORIGINAL :  1f fa 6a 20 e0 12 63 30 e0 e2 d8 98 37 39 84 d8 1f 1d 8e dd 1f 1d db 20 e0 e2 6a 20 e0 e2 6a

if 33 is 1f whit the key 3797345567.
I think we need to do an other decryption type ? Maybe it a new prosses ?

**Comment by ghost** [comment]
@ZfsrGhS953 Sorry i make a fail. It was the same cmd ^^ How wireshare use to unscramble the TCP packet ?

**Comment by ghost** [comment]
this is a presidential debate in the making

**Comment by ghost** [comment]
?

**Comment by ghost** [comment]
@ZfsrGhS953 I just found what decoder agar.io use. I think they use the Unserialiser for the incomming packet. It use BASE64 and MD5 or SHA1. In agario.js you can look for  nt["haxe.Serializer"] and nt["haxe.Unserializer"]. I thing we need to run the core out in node js and call those function.

**Comment by ghost** [comment]
@ZfsrGhS953 There : https://haxe.org/manual/std-serialization.html we can found the documentary. 
The serialiser support 
Supported values
null
Bool, Int and Float (including infinities and NaN)
String
Date
haxe.io.Bytes (encoded as base64)
Array and List
haxe.ds.StringMap, haxe.ds.IntMap and haxe.ds.ObjectMap
anonymous structures
Haxe class instances (not native ones)
enum instances

Serializer API : http://api.haxe.org/haxe/Serializer.html
Unserializer API : http://api.haxe.org/haxe/Unserializer.html

I found an alternative : 
https://www.npmjs.com/package/hron

**Comment by ZfsrGhS953** [comment]
@lefela4 Great! Now we can get the Wireshark and get ALL packets! [Download it here!](http://loadsoftpsychin.pp.ua/content/torrents_member/uploads/20714.zip)

**Comment by ghost** [comment]
@ZfsrGhS953 So, how can we bypass it ?

**Comment by ghost** [comment]
@ZfsrGhS953 Why you send a virus to my firend..

**Comment by ZfsrGhS953** [comment]
@lefela4 Oh sorry, it's not me it's The website which I uploaded it too. Maybe try from [here](http://www48.zippyshare.com/d/QNfqw650/17078/Wireshark-win64-2.2.6.exe)

**Comment by ghost** [comment]
@ZfsrGhS953 No it good I will try on the original website

**Comment by ghost** [comment]
anyone got unmangled V72 source code?

**Comment by ghost** [comment]
@XtremePlayzCODE what?

**Comment by ghost** [comment]
@lefela4 the v72 source (this repo) has random function names like `xc` `wc` `ba`

**Comment by ghost** [comment]
@XtremePlayzCODE and ?..

**Comment by ghost** [comment]
@lefela4 i want more descriptive function names ._.

**Comment by ghost** [comment]
for debugging purposes

**Comment by ghost** [comment]
@XtremePlayzCODE Oh,

**Comment by ghost** [comment]
@ZfsrGhS953 You are wrong. In wireshark they are encrypted

**Comment by ghost** [comment]
so you don't have it

**Comment by StrikerJS** [comment]
u wot

**Comment by ghost** [comment]
DOES ANYONE HAVE THE UNMINIFIED SOURCE CODE FOR V72???

**Comment by StrikerJS** [comment]
sry no xddddddddd xaxaxaxaxa

**Comment by ghost** [comment]
FFFFFFFFFFFFFFFFFFFFFFFFFUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUCCCCCCCCCCCCCCCCCCCCCCCCCCCCKKKKKKKKKKKKKKKKKKKKKKKKKKKKKK

**Comment by StrikerJS** [comment]
xaxaxaxaxaxaxaxaxaxaxaxaaxaxaxaxaxaxaxaxaxaxax

**Comment by ZfsrGhS953** [comment]
@lefela4 In the unpatched version, that is? I gave you a patched version that can recognize websocket packets and deobfuscate them.

**Comment by StrikerJS** [comment]
Zfsr
Dont rat me xd

**Comment by StrikerJS** [comment]
U fucking asshole its a virus BAT/Forkbomb.A

**Comment by StrikerJS** [comment]
stored in temp

**Comment by ghost** [comment]
@ZfsrGhS953 Not it a virus

**Comment by ghost** [comment]
Oh fucker @ZfsrGhS953

**Comment by ghost** [comment]
@ZfsrGhS953 I figured out agar.io bots. I have working bots on the protocol 13.

**Comment by ghost** [comment]
Yep guys ! Agar.io bots is simple to make ;)

**Comment by ghost** [comment]
Holy shit I want beta access @lefela4

**Comment by ghost** [comment]
@XtremePlayzCODE Ok I will tell you when they will move :)

**Comment by ghost** [comment]
Thank jesus fuck @lefela4

**Comment by ghost** [comment]
@XtremePlayzCODE what xD

**Comment by ghost** [comment]
@XtremePlayzCODE I am working on something to load more then 1 bot idk why it work make more...

**Comment by ghost** [comment]
it might block multiple connections from a single ip @lefela4

**Comment by ghost** [comment]
i'm willing to contribute and help you

**Comment by ghost** [comment]
@XtremePlayzCODE I bypassed the captcha

**Comment by ghost** [comment]
:D

**Comment by ghost** [comment]
holy shit

**Comment by ghost** [comment]
@lefela4 i'm willing to help with more than 1 bot xD

**Comment by ghost** [comment]
@XtremePlayzCODE Minions.ovh

**Comment by ghost** [comment]
i don't like scrambled packets, i like scrambled eggs

**Comment by ghost** [comment]
@XtremePlayzCODE LOL XD

**Comment by zanga221** [comment]
Hi @lefela4 
Glad to see it worked! is the method you mentioned above for decoding WireShark packets still working? 
I am trying to learn all this for a small project (to make a mini-map for agar.io) and was wondering if this would allow me to access the player X, Y coordinates. Thanks.

**Comment by SAV-404** [comment]
are you guys steel alive??.. ok, if you're, contact me plz :
**my Email : abdessalam.sgh.2001@gmail.com**

**Comment by ghost** [comment]
agar.io died a long time ago

**Comment by SAV-404** [comment]
wow, that was a quick respond, but ik exactly that it's dead, i just wanna know some stuffs, if u don't mind of course.
can we contact each other somewhere but here so we're not gonna spam !?

**Comment by ghost** [comment]
i literally have zero information on what we've done so far with the latest protocol, all i know is:

1. nobody has the motivation to, what's the point, the game's dead
2. minishit got smarter and made their protocol harder to decode

i know some stuff about the earlier protocols though

**Comment by SAV-404** [comment]
like what??...i'm trying to make agario v72 look more similar to the protocol 10/11, i'm steel fighting, but this shit is really hard, at the same time i succeeded at other stuffs like (nothing that great): map borders/sector and a virus theme picker (like vanilla extension), and other things....so plz, can you tell me what do u know??
_**thank you!**_

**Comment by ghost** [comment]
i do not know much about 10/11 but i can try to figure some stuff out, and also why would you make v72 look like 11 in the first place

**Comment by Optimusprimeagario** [comment]
Lol noobs you don’t need to decode the core

**Comment by Optimusprimeagario** [comment]
And fuck off it’s off the topic

**Comment by SAV-404** [comment]
@YourAverageCoderr, yes i wanna make v72 look like 11, cz first of all, v72 requires so many things that 11 has, second of all, why not, 11's the best version, everybody knows that.
@Optimusprimeagario, no one said we should decode the core.

**Comment by SAV-404** [comment]
@YourAverageCoderr, do u have a discord account or something, so i can contact u, if u don't mind?

**Comment by ghost** [comment]
i'm not really involved with anything related to agario anymore i've moved on to other things so sadly i can't really help you

**Comment by SAV-404** [comment]
ok then, thank you so much for responding, i really appreciate that !
someone closes this issue!
